### PR TITLE
arch(rpc-decode): drain the single-level positional allowlist to zero (19 files, #1491/#1501)

### PR DESCRIPTION
**Pull request by teng-lin**
## What

Completes the `#1491` / `#1501` burndown: migrates **every** below-facade `name[int]` positional read of a decoded `batchexecute` RPC payload behind `safe_index()` / typed row adapters, and **empties `SINGLE_LEVEL_ALLOWLIST` (19 files → 0)**. Both positional gates (chained `ALLOWLIST` and single-level `SINGLE_LEVEL_ALLOWLIST`) are now empty, so the whole feature tree is hard-protected: any *new* positional RPC read fails CI.

123 raw reads across 19 source files, drained by adapter domain:

| Domain | Files | How |
|---|---|---|
| artifacts | formatters, polling, downloads, listing, `_artifacts.py` | `unwrap_artifact_rows` / `unwrap_mind_map_generation_leaf` + guarded `safe_index`; several were type-blind false positives (typed `Artifact`/exception lists) restructured to `next(iter(...))` |
| sources | `_source/{listing,add,upload,content}.py`, `_types/sources.py` | new `SourceGuideRow`/`SourceFulltextRow`, `SourceRow.url_from_metadata`/`created_at_from_metadata` |
| notes | `_notebooks.py`, `_note_service.py`, `_types/notebooks.py` | guarded inline `safe_index` |
| chat+labels | `_chat/api.py`, `_chat/notes.py`, `_labels.py` | new `unwrap_last_conversation_id` + `SavedChatNoteRow` |
| research+mind_maps | `_research.py`, `_mind_maps_api.py` | new `ResearchStartRow`/`ImportedSourceRow`/`unwrap_import_rows`; mind_maps via guarded inline `safe_index` |
| sharing | `_types/sharing.py` | guarded inline `safe_index` |

## Why it's safe (behavior preservation)

The invariant was **exact behavior preservation** — RPC decode is the repo's #1 breakage class. The mechanics:
- **Soft reads stay soft.** Where the legacy code returned `None`/`[]`/`""`/skip on missing/short/malformed input, the `safe_index` is placed *after* the guard that proves the slot present, so the strict helper is a no-op on every reachable input and the soft contract is preserved. No soft return-path was converted to a raise.
- **`next(iter(x))` only where `x` is provably non-empty** (guarded by `if not x`/`len(x)==1`/`if x:` directly above), so no `IndexError`→`StopIteration` drift.
- **`except` tuples that wrapped `safe_index` had `UnknownRPCMethodError` added** where drift could surface, so swallow-paths behave identically.
- **Differential parity tests**: each domain ships tests that embed verbatim copies of the pre-drain logic as oracles and assert byte-for-byte parity across present / empty / too-short / malformed / `None` / scalar inputs.

**Net robustness gain:** genuine schema drift now **fails loud** (`UnknownRPCMethodError` with diagnostics) instead of silently swallowing to a wrong value — the ADR-0011 goal.

## One intentional, documented deviation

`_notebooks.py::get_source_ids` now routes id decode through `SourceRow.id`, which **stringifies** a non-string source id, where the legacy path **silently dropped** non-string ids. This is unobservable on today's wire (Google never sends non-string source ids), aligns the path with its sibling `_source.listing._extract_source_id`, and is covered by tests. Flagging explicitly so reviewers can object — happy to revert to a strict drop if preferred.

## Review & verification

- **Polish gate run** (claude + codex, no agy): Stage 1 conservative simplify, Stage 2 dual behavior-preservation review — **both reviewers found zero Critical/Important defects**; the two Minor test-coverage gaps codex raised are closed in this PR.
- Full suite green, `mypy` clean (271 files), `ruff` clean, all 1055 guardrail tests pass (incl. the now-empty single-level gate + its stale-allowlist self-check).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/teng-lin/notebooklm-py/pull/1557?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Mind‑map generation now reliably creates and saves notes when a mind‑map is returned.

* **Refactor**
  * Standardized server-response parsing across artifacts, chat, research, sources, sharing, and notebooks for more predictable behavior.

* **Bug Fixes**
  * More robust selection, ordering, and error propagation for artifact downloads, listings, labels, uploads, notes, and chat-related flows.

* **Tests**
  * Expanded unit tests covering parsing, unwrapping, and malformed-edge cases to prevent regressions.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_artifact/downloads.py [hunk 1]
```diff
@@ -468,7 +468,7 @@ async def download_interactive_artifact(
             if not artifact:
                 raise ArtifactNotFoundError(artifact_id, artifact_type=artifact_type)
         else:
-            artifact = completed[0]
+            artifact = next(iter(completed))  # typed ``Artifact`` list head (newest-first)
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_artifact/downloads.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_artifact/listing.py [hunk 2]
```diff
@@ -290,7 +290,10 @@ def select_completed_artifact_row(
         # the ``test_handles_none_at_timestamp_position_without_typeerror``
         # contract).
         filtered.sort(key=lambda row: row.created_at_raw or 0, reverse=True)
-        return filtered[0]
+        # ``filtered`` is a non-empty list of typed ``ArtifactRow`` objects (not
+        # a raw RPC payload); take the most-recent head via ``next(iter(...))``
+        # so this typed-sequence pick is not the ``name[int]`` RPC-row shape.
+        return next(iter(filtered))
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_artifact/listing.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_labels.py [hunk 3]
```diff
@@ -186,7 +186,12 @@ async def create(self, notebook_id: str, name: str, emoji: str = "") -> Label:
                 f"create(name={name!r}) expected exactly 1 new label, found {len(new)} "
                 f"(concurrent label creation can cause this — retry from a fresh list)"
             )
-        return new[0]
+        # ``new`` is a list[Label] (typed dataclass instances), not a decoded
+        # RPC payload — positional RPC-row decode already happened in
+        # ``_labels_from_envelope``/``LabelRow``. ``next(iter(...))`` avoids the
+        # type-blind single-level ``name[int]`` guardrail false-positive that a
+        # ``new[0]`` index would trip.
+        return next(iter(new))
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_labels.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_notebooks.py [hunk 4]
```diff
@@ -433,7 +476,11 @@ async def _probe() -> Notebook | None:
                 return None
             matches = [nb for nb in current if nb.id not in baseline_ids and nb.title == title]
             if len(matches) == 1:
-                return matches[0]
+                # ``matches`` is a list of typed ``Notebook`` objects (NOT a raw
+                # RPC payload) — ``next(iter(...))`` reads the single match
+                # without the ``name[int]`` shape that the positional-decode gate
+                # (rightly) flags only for genuine payload descents.
+                return next(iter(matches))
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_notebooks.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_source/upload.py [hunk 5]
```diff
@@ -854,7 +857,7 @@ async def _probe() -> str | None:
                     ),
                 )
             if len(matches) == 1:
-                return matches[0].id
+                return next(iter(matches)).id  # not ``matches[0]``: ratchet
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_source/upload.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_source/upload.py [hunk 6]
```diff
@@ -282,13 +282,13 @@ def _extract_register_file_source_id(result: Any, filename: str) -> str | None:
     """
     field_candidates = _extract_source_id_field_candidates(result, filename)
     if len(field_candidates) == 1:
-        return field_candidates[0]
+        return next(iter(field_candidates))  # next(iter): type-blind ratchet
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_source/upload.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_source/upload.py [hunk 7]
```diff
@@ -282,13 +282,13 @@ def _extract_register_file_source_id(result: Any, filename: str) -> str | None:
...
     if len(field_candidates) > 1:
         return None
 
     row_candidates = _extract_contextual_source_id_row_candidates(result, filename)
     if len(row_candidates) == 1:
-        return row_candidates[0]
+        return next(iter(row_candidates))
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_source/upload.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_research.py [hunk 8]
```diff
@@ -456,7 +458,9 @@ async def poll(
         )
 
         if parsed_tasks:
-            return self._public_poll_result(parsed_tasks[0], parsed_tasks)
+            # ``parsed_tasks`` is a typed ``list[ResearchTask]`` (not a decoded
+            # RPC payload); ``next(iter(...))`` takes the first without a subscript.
+            return self._public_poll_result(next(iter(parsed_tasks)), parsed_tasks)
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_research.py]: Addressed in 91d9a4a6 — switched to sequence unpacking (`(x,) = …` for the `len == 1`-guarded picks, `first, *_ = …` for the non-empty first-of-many picks). More idiomatic, makes the single-element invariant explicit, and sidesteps the PEP 479 `StopIteration`→`RuntimeError` nuance you flagged. Still avoids the type-blind `name[int]` positional-decode ratchet, and behavior is unchanged (every site is guarded non-empty / exactly-one).

File: src/notebooklm/_row_adapters/research.py [hunk 9]
```diff
@@ -301,3 +304,133 @@ def deep_payload(payload: Any) -> tuple[str, str] | None:
...
+    if not result or not isinstance(result, list):
+        return []
+    first = result[_IMPORT_ENVELOPE_OUTER_POS]
+    if (
+        isinstance(first, list)
+        and len(first) > 0
+        and isinstance(first[_IMPORT_ENVELOPE_PROBE_POS], list)
+    ):
+        return first
+    return result
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_row_adapters/research.py:385]: Verified against `origin/main`: this is **faithful preservation of the pre-existing inline heuristic, not a regression**. Legacy `import_sources` ran the identical probe — `if len(result)>0 and isinstance(result[0],list) and len(result[0])>0: first=result[0]; if isinstance(first[0],list): result=first` — so for `[[["id"],"t"]]` it likewise rebinds `result` to `[["id"],"t"]`, iterates `["id"]` and `"t"` as rows, and skips both → `[]`. `unwrap_import_rows` reproduces this exactly (pinned by a parity test in test_research_row_adapter.py). Since #1557 is a strictly behavior-preserving decode drain, fixing the ambiguity here would smuggle a behavior change into a refactor — so I've filed #1558 to disambiguate the IMPORT_RESEARCH envelope (needs a golden capture to confirm reachability). Thanks for the catch.

**Comment by teng-lin** [comment]
Addressed the out-of-diff stale-comment finding in 571d3476: the block-comments in `_labels.py`, `_artifact/listing.py`, `_notebooks.py`, and `_research.py` now describe the actual tuple-unpacking pattern instead of the old `next(iter(...))`. Comment-only change; `_research.py` stays at its 937-LOC ceiling.
